internal/reconciler/certificate/reconciler.go

```diff
@@ -0,0 +1,240 @@
+		}
+		return reconcile.Result{}, err
+	}
+
+	if project.Spec.Namespace == nil {
+		log.Info("Removing certificates from store - project spec.namespace is nil", "project", projectName)
+		r.deleteMapping(mappingKey)
+
+		return reconcile.Result{}, nil
+	}
+
+	if *project.Spec.Namespace != namespace.Name {
+		log.Info("Removing certificates from store - namespace name does not match project namespace", "project", projectName)
+		r.deleteMapping(mappingKey)
+
+		return reconcile.Result{}, nil
+	}
+
+	shootName, ok = configmap.Labels[v1beta1constants.LabelShootName]
+	if !ok {
+		log.Info("Removing certificates from store - configmap does not have expected label", "label", v1beta1constants.LabelShootName)
+		r.deleteMapping(mappingKey)
+		return reconcile.Result{}, nil
+	}
+
+	shootUID, ok = configmap.Labels["shoot.gardener.cloud/uid"]
+	if !ok {
+		log.Info("Removing certificates from store - configmap does not have expected label", "label", "shoot.gardener.cloud/uid")
+		r.deleteMapping(mappingKey)
+		return reconcile.Result{}, nil
+	}
+
+	shoot := &gardencorev1beta1.Shoot{ObjectMeta: metav1.ObjectMeta{
+		Name:      shootName,
+		Namespace: req.Namespace,
+	}}
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(shoot), shoot); err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Info("Removing certificates from store - shoot not found", "shoot", shoot)
+			r.deleteMapping(mappingKey)
+
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	if shootUID != string(shoot.UID) {
+		log.Info("Removing certificates from store - shoot UID is different in spec and configmap label", "shoot", shoot)
+		r.deleteMapping(mappingKey)
+		return reconcile.Result{}, nil
+	}
+
+	data := configmap.Data[secretsutils.DataKeyCertificateCA]
+
+	certs := []byte(data)
+	for {
+		block, rest := pem.Decode(certs)
+		if block == nil {
+			break
+		}
+
+		if block.Type != "CERTIFICATE" {
+			log.Info("Removing certificates from store - block type is not CERTIFICATE")
+			r.deleteMapping(mappingKey)
+			return reconcile.Result{}, nil
+		}
+
+		if len(block.Headers) > 0 {
+			log.Info("Removing certificates from store - block headers are not expected")
+			r.deleteMapping(mappingKey)
+			return reconcile.Result{}, nil
+		}
+
+		cert, err := x509.ParseCertificate(block.Bytes)
+		if err != nil {
+			// we cannot do anything until this is fixed
+			log.Error(err, "Removing certificates from store - failed to parse certificate")
+			r.deleteMapping(mappingKey)
+			return reconcile.Result{}, nil
+		}
+
+		if !cert.IsCA {
+			log.Info("Removing certificates from store - certificate is not a CA")
+			r.deleteMapping(mappingKey)
+			return reconcile.Result{}, nil
+		}
+
+		certs = rest
+	}
+
+	// Finally write the certificates to store
+	log.Info("Adding certificates to store", "shoot", shoot)
+	bundle := struct {
+		Certs string `json:"certs"`
+	}{Certs: data}
+
+	payload, err := json.Marshal(bundle)
+	if err != nil {
+		return ctrl.Result{}, err
```

Comment:
Why is the certificate not removed from the store in this case?

Reply:
I probably missed that. I remember trying to figure out when can this fail. I suppose that we would not see many failures on that path.